velox/dwio/dwrf/reader/SelectiveColumnReader.cpp

```diff
@@ -74,27 +74,30 @@ SelectiveColumnReader::SelectiveColumnReader(
     const EncodingKey& ek,
     StripeStreams& stripe,
     common::ScanSpec* scanSpec,
-    const TypePtr& type,
-    bool loadIndex)
+    const TypePtr& type)
     : ColumnReader(ek, stripe),
       scanSpec_(scanSpec),
       type_{type},
       rowsPerRowGroup_{stripe.rowsPerRowGroup()} {
-  if (scanSpec->filter() || loadIndex) {
-    indexStream_ =
-        stripe.getStream(ek.forKind(proto::Stream_Kind_ROW_INDEX), false);
-  }
+  // We always initialize indexStream_ because indices are needed as
+  // soon as there is a single filter that can trigger row group skips
+  // anywhere in the reader tree. This is not known at construct time
+  // because the first filter can come from a hash join or other run
+  // time pushdown.
+  indexStream_ =
+      stripe.getStream(ek.forKind(proto::Stream_Kind_ROW_INDEX), false);
 }
 
 std::vector<uint32_t> SelectiveColumnReader::filterRowGroups(
     uint64_t rowGroupSize,
     const StatsContext& context) const {
-  ensureRowGroupIndex();
-  auto filter = scanSpec_->filter();
-  if (!index_ || !filter) {
+  if ((!index_ && !indexStream_) || !scanSpec_->filter()) {
```

Comment:
is !index_ && !indexStream_ a valid case? Looks to me we will always have one of them